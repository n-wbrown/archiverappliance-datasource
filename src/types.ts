import { DataQuery, DataSourceJsonData } from '@grafana/data';

export interface AAQuery extends DataQuery {
  target: string;
  alias: string;
  operator: string;
  regex: boolean;
  aliasPattern: string;
  functions: any[];
}

export const defaultQuery: Partial<AAQuery> = {
  target: '',
  alias: '',
  operator: '',
  regex: false,
  aliasPattern: '',
  functions: [],
};

/**
 * These are options configured for each DataSource instance
 */
export interface AADataSourceOptions extends DataSourceJsonData {
  path?: string;
}

/**
 * Value that is used in the backend, but never sent over HTTP to the frontend
 */
export interface AASecureJsonData {
  apiKey?: string;
}
