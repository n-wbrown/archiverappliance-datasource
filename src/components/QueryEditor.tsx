import defaults from 'lodash/defaults';
import React, { ChangeEvent, PureComponent, KeyboardEvent } from 'react';
import { InlineFormLabel, LegacyForms } from '@grafana/ui';
import { QueryEditorProps } from '@grafana/data';
import { getTemplateSrv } from '@grafana/runtime';
import Autosuggest from 'react-autosuggest';
import { DataSource } from '../DataSource';
import { AADataSourceOptions, AAQuery, defaultQuery, operatorList, FunctionDescriptor } from '../types';

import { Functions } from './Functions';

type Props = QueryEditorProps<DataSource, AAQuery, AADataSourceOptions>;

interface State {
  pvSuggestions: any[];
  oprSuggestions: any[];
}

const colorYellow = '#d69e2e';

const getSuggestionValue = (suggestion: any) => {
  return suggestion;
};

const renderSuggestion = (suggestion: any) => {
  return <span>{suggestion}</span>;
};

export class QueryEditor extends PureComponent<Props, State> {
  state = { pvSuggestions: [], oprSuggestions: [] };

  onPVChange = (event: ChangeEvent<HTMLInputElement>, { newValue }: any) => {
    const { onChange, query } = this.props;
    onChange({ ...query, target: newValue });
  };

  onRegexChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const { onChange, onRunQuery, query } = this.props;
    onChange({ ...query, regex: !query.regex });
    onRunQuery();
  };

  onOperatorChange = (event: ChangeEvent<HTMLInputElement>, { newValue }: any) => {
    const { onChange, query } = this.props;
    onChange({ ...query, operator: newValue });
  };

  onAliasChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { onChange, query } = this.props;
    onChange({ ...query, alias: event.target.value });
  };

  onAliaspatternChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { onChange, query } = this.props;
    onChange({ ...query, aliasPattern: event.target.value });
  };

  onFuncsChange = (funcs: FunctionDescriptor[]) => {
    const { onChange, query } = this.props;
    onChange({ ...query, functions: funcs });
  };

  onKeydownEnter = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.keyCode === 13) {
      event.currentTarget.blur();
    }
  };

  loadPVSuggestions(value: string) {
    const templateSrv = getTemplateSrv();
    const replacedQuery = templateSrv.replace(value, undefined, 'regex');
    const { regex } = this.props.query;
    const searchQuery = regex ? replacedQuery : `.*${replacedQuery}.*`;
    this.props.datasource.pvNamesFindQuery(searchQuery, 100).then((res: any) => {
      this.setState({
        pvSuggestions: res,
      });
    });
  }

  onPVSuggestionsFetchRequested = ({ value }: { value: any }) => {
    this.loadPVSuggestions(value);
  };

  onPVSuggestionsClearRequested = () => {
    this.setState({
      pvSuggestions: [],
    });
  };

  onOprSuggestionsFetchRequested = ({ value }: { value: any }) => {
    const escapedValue = value.trim().replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const regex = new RegExp('.*' + escapedValue, 'i');

    const suggestions = operatorList.filter(operator => regex.test(operator));

    this.setState({
      oprSuggestions: suggestions,
    });
  };

  onOprSuggestionsClearRequested = () => {
    this.setState({
      oprSuggestions: [],
    });
  };

  onSuggestionsSelected = (
    event: React.FormEvent<any>,
    {
      suggestion,
      suggestionValue,
      suggestionIndex,
      sectionIndex,
      method,
    }: {
      suggestion: any;
      suggestionValue: string;
      suggestionIndex: number;
      sectionIndex: number | null;
      method: 'click' | 'enter';
    }
  ) => {
    if (method === 'enter') {
      event.currentTarget.blur();
    }
  };

  render() {
    const query = defaults(this.props.query, defaultQuery);
    const { pvSuggestions, oprSuggestions } = this.state;
    const { onRunQuery } = this.props;
    const pvInputStyle = query.regex ? { color: colorYellow } : {};
    const aliasInputStyle = query.aliasPattern ? { color: colorYellow } : {};

    return (
      <>
        <div className="gf-form-inline">
          <InlineFormLabel width={6} className="query-keyword">
            PV
          </InlineFormLabel>
          <div className="max-width-30" style={{ marginRight: '4px' }}>
            <Autosuggest
              suggestions={pvSuggestions}
              onSuggestionsFetchRequested={this.onPVSuggestionsFetchRequested}
              onSuggestionsClearRequested={this.onPVSuggestionsClearRequested}
              getSuggestionValue={getSuggestionValue}
              onSuggestionSelected={this.onSuggestionsSelected}
              renderSuggestion={renderSuggestion}
              shouldRenderSuggestions={() => true}
              inputProps={{
                value: query.target,
                className: 'gf-form-input',
                placeholder: 'PV name',
                spellCheck: false,
                style: pvInputStyle,
                onChange: this.onPVChange,
                onBlur: onRunQuery,
                onKeyDown: this.onKeydownEnter,
              }}
            />
          </div>
          <LegacyForms.Switch
            checked={query.regex}
            label="Regex"
            labelClass={'width-6  query-keyword'}
            onChange={this.onRegexChange}
          />
        </div>
        <div className="gf-form">
          <InlineFormLabel width={6} className="query-keyword">
            Operator
          </InlineFormLabel>
          <div className="max-width-30">
            <Autosuggest
              suggestions={oprSuggestions}
              onSuggestionsFetchRequested={this.onOprSuggestionsFetchRequested}
              onSuggestionsClearRequested={this.onOprSuggestionsClearRequested}
              onSuggestionSelected={this.onSuggestionsSelected}
              getSuggestionValue={getSuggestionValue}
              renderSuggestion={renderSuggestion}
              shouldRenderSuggestions={() => true}
              inputProps={{
                value: query.operator,
                className: 'gf-form-input',
                placeholder: 'mean',
                spellCheck: false,
                onChange: this.onOperatorChange,
                onBlur: onRunQuery,
                onKeyDown: this.onKeydownEnter,
              }}
            />
          </div>
        </div>
        <div className="gf-form">
          <InlineFormLabel width={6} className="query-keyword">
            Alias
          </InlineFormLabel>
          <input
            type="text"
            value={query.alias}
            className="gf-form-input max-width-30"
            placeholder="Alias"
            style={aliasInputStyle}
            onChange={this.onAliasChange}
            onBlur={onRunQuery}
            onKeyDown={this.onKeydownEnter}
          />
          <InlineFormLabel width={6} className="query-keyword">
            Alias pattern
          </InlineFormLabel>
          <input
            type="text"
            value={query.aliasPattern}
            className="gf-form-input max-width-30"
            placeholder="Alias regex pattern"
            style={{ color: colorYellow }}
            onChange={this.onAliaspatternChange}
            onBlur={onRunQuery}
            onKeyDown={this.onKeydownEnter}
          />
        </div>
        <Functions funcs={query.functions} onChange={this.onFuncsChange} onRunQuery={onRunQuery} />
      </>
    );
  }
}
