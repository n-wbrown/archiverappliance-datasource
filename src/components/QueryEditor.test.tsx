import defaults from 'lodash/defaults';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { QueryEditor } from './QueryEditor';
import { DataSource } from '../DataSource';
import { AAQuery, defaultQuery } from '../types';

const setup = (propOverrides?: object) => {
  const datasourceMock: unknown = {
    createQuery: jest.fn(q => q),
    pvNamesFindQuery: jest.fn((q, num) => ['PV']),
  };
  const datasource: DataSource = datasourceMock as DataSource;
  const onRunQuery = jest.fn();
  const onChange = jest.fn();
  const query: AAQuery = {
    target: 'pvname',
    alias: 'alias',
    operator: 'mean',
    regex: false,
    aliasPattern: '',
    functions: [],
    refId: 'A',
  };

  const props: any = {
    datasource,
    onChange,
    onRunQuery,
    query,
  };

  Object.assign(props, propOverrides);

  const wrapper = shallow(<QueryEditor {...props} />);
  const instance = wrapper.instance() as QueryEditor;

  return {
    instance,
    wrapper,
  };
};

describe('Render Editor with basic options', () => {
  it('should render normally', () => {
    const { wrapper } = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render regex mode', () => {
    const props = defaults({ regex: true }, defaultQuery);
    const { wrapper } = setup({ query: props });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render with alias pattern', () => {
    const props = defaults({ aliasPattern: '.*' }, defaultQuery);
    const { wrapper } = setup({ query: props });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render with top function', () => {
    const func = [
      {
        def: {
          category: 'Filter Series',
          defaultParams: [5, 'avg'],
          name: 'top',
          params: [
            {
              name: 'number',
              type: 'int',
            },
            {
              name: 'value',
              options: ['avg', 'min', 'max', 'absoluteMin', 'absoluteMax', 'sum'],
              type: 'string',
            },
          ],
        },
        params: ['5', 'avg'],
        text: 'top(5, avg)',
      },
    ];
    const props = defaults({ functions: func }, defaultQuery);
    const { wrapper } = setup({ query: props });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
