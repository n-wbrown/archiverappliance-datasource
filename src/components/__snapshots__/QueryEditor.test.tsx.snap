// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Editor with basic options should render normally 1`] = `
<Fragment>
  <div
    className="gf-form-inline"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      PV
    </FormLabel>
    <div
      className="max-width-30"
      style={
        Object {
          "marginRight": "4px",
        }
      }
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "PV name",
            "spellCheck": false,
            "style": Object {},
            "value": "pvname",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
    <Switch
      checked={false}
      label="Regex"
      labelClass="width-6  query-keyword"
      onChange={[Function]}
    />
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Operator
    </FormLabel>
    <div
      className="max-width-30"
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "mean",
            "spellCheck": false,
            "value": "mean",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias"
      style={Object {}}
      type="text"
      value="alias"
    />
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias pattern
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias regex pattern"
      style={
        Object {
          "color": "#d69e2e",
        }
      }
      type="text"
      value=""
    />
  </div>
  <Functions
    funcs={Array []}
    onChange={[Function]}
    onRunQuery={[MockFunction]}
  />
</Fragment>
`;

exports[`Render Editor with basic options should render regex mode 1`] = `
<Fragment>
  <div
    className="gf-form-inline"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      PV
    </FormLabel>
    <div
      className="max-width-30"
      style={
        Object {
          "marginRight": "4px",
        }
      }
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "PV name",
            "spellCheck": false,
            "style": Object {
              "color": "#d69e2e",
            },
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
    <Switch
      checked={true}
      label="Regex"
      labelClass="width-6  query-keyword"
      onChange={[Function]}
    />
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Operator
    </FormLabel>
    <div
      className="max-width-30"
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "mean",
            "spellCheck": false,
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias"
      style={Object {}}
      type="text"
      value=""
    />
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias pattern
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias regex pattern"
      style={
        Object {
          "color": "#d69e2e",
        }
      }
      type="text"
      value=""
    />
  </div>
  <Functions
    funcs={Array []}
    onChange={[Function]}
    onRunQuery={[MockFunction]}
  />
</Fragment>
`;

exports[`Render Editor with basic options should render with alias pattern 1`] = `
<Fragment>
  <div
    className="gf-form-inline"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      PV
    </FormLabel>
    <div
      className="max-width-30"
      style={
        Object {
          "marginRight": "4px",
        }
      }
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "PV name",
            "spellCheck": false,
            "style": Object {},
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
    <Switch
      checked={false}
      label="Regex"
      labelClass="width-6  query-keyword"
      onChange={[Function]}
    />
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Operator
    </FormLabel>
    <div
      className="max-width-30"
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "mean",
            "spellCheck": false,
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias"
      style={
        Object {
          "color": "#d69e2e",
        }
      }
      type="text"
      value=""
    />
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias pattern
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias regex pattern"
      style={
        Object {
          "color": "#d69e2e",
        }
      }
      type="text"
      value=".*"
    />
  </div>
  <Functions
    funcs={Array []}
    onChange={[Function]}
    onRunQuery={[MockFunction]}
  />
</Fragment>
`;

exports[`Render Editor with basic options should render with top function 1`] = `
<Fragment>
  <div
    className="gf-form-inline"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      PV
    </FormLabel>
    <div
      className="max-width-30"
      style={
        Object {
          "marginRight": "4px",
        }
      }
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "PV name",
            "spellCheck": false,
            "style": Object {},
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
    <Switch
      checked={false}
      label="Regex"
      labelClass="width-6  query-keyword"
      onChange={[Function]}
    />
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Operator
    </FormLabel>
    <div
      className="max-width-30"
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "className": "gf-form-input",
            "onBlur": [MockFunction],
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "mean",
            "spellCheck": false,
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "react-autosuggest__container",
            "containerOpen": "react-autosuggest__container--open",
            "input": "react-autosuggest__input",
            "inputFocused": "react-autosuggest__input--focused",
            "inputOpen": "react-autosuggest__input--open",
            "sectionContainer": "react-autosuggest__section-container",
            "sectionContainerFirst": "react-autosuggest__section-container--first",
            "sectionTitle": "react-autosuggest__section-title",
            "suggestion": "react-autosuggest__suggestion",
            "suggestionFirst": "react-autosuggest__suggestion--first",
            "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
            "suggestionsContainer": "react-autosuggest__suggestions-container",
            "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
            "suggestionsList": "react-autosuggest__suggestions-list",
          }
        }
      />
    </div>
  </div>
  <div
    className="gf-form"
  >
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias"
      style={Object {}}
      type="text"
      value=""
    />
    <FormLabel
      className="query-keyword"
      width={6}
    >
      Alias pattern
    </FormLabel>
    <input
      className="gf-form-input max-width-30"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Alias regex pattern"
      style={
        Object {
          "color": "#d69e2e",
        }
      }
      type="text"
      value=""
    />
  </div>
  <Functions
    funcs={
      Array [
        Object {
          "def": Object {
            "category": "Filter Series",
            "defaultParams": Array [
              5,
              "avg",
            ],
            "name": "top",
            "params": Array [
              Object {
                "name": "number",
                "type": "int",
              },
              Object {
                "name": "value",
                "options": Array [
                  "avg",
                  "min",
                  "max",
                  "absoluteMin",
                  "absoluteMax",
                  "sum",
                ],
                "type": "string",
              },
            ],
          },
          "params": Array [
            "5",
            "avg",
          ],
          "text": "top(5, avg)",
        },
      ]
    }
    onChange={[Function]}
    onRunQuery={[MockFunction]}
  />
</Fragment>
`;
